src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,621 @@
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.TrafficWatch;
+using Sparrow;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedBatchHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/bulk_docs", "POST")]
+        public async Task BulkDocs()
+        {
+            var commandBuilder = new ShardedBatchCommandBuilder(this);
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var contentType = HttpContext.Request.ContentType;
+                if (contentType == null ||
+                    contentType.StartsWith("application/json", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.BuildCommandsAsync(context, RequestBodyStream());
+                }
+                else if (contentType.StartsWith("multipart/mixed", StringComparison.OrdinalIgnoreCase) ||
+                         contentType.StartsWith("multipart/form-data", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.ParseMultipart(context, RequestBodyStream(), HttpContext.Request.ContentType);
+                }
+                else
+                    BatchHandler.ThrowNotSupportedType(contentType);
+
+                if (TrafficWatchManager.HasRegisteredClients)
+                {
+                    var log = BatchHandler.BatchTrafficWatch(commandBuilder.Commands);
+                    // add sb to httpContext
+                    AddStringToHttpContext(log, TrafficWatchChangeType.BulkDocs); // TODO sharding: maybe ShardedBulkDocs type?
+                }
+
+                using (var batch = await commandBuilder.GetCommand(context))
+                {
+                    if (batch.IsClusterTransaction)
+                    {
+                        throw new NotSupportedException("Cluster transactions are currently not supported in a sharded database.");
+                    }
+
+                    var shardedBatchCommands = new Dictionary<int, SingleNodeShardedBatchCommand>(); // TODO sharding : consider cache those
+                    foreach (var command in batch)
+                    {
+                        var shardIndex = command.Shard;
+                        var requestExecutor = ShardedContext.RequestExecutors[shardIndex];
+
+                        if (shardedBatchCommands.TryGetValue(shardIndex, out var shardedBatchCommand) == false)
+                        {
+                            shardedBatchCommand = new SingleNodeShardedBatchCommand(this, requestExecutor.ContextPool);
+                            shardedBatchCommands.Add(shardIndex, shardedBatchCommand);
+                        }
+
+                        shardedBatchCommand.AddCommand(command);
+                    }
+
```

Comment:
Let's create an issue here to add an option to reject multi shard transactions

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-16276